db/skipped_sequence.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2016-Present Couchbase, Inc.
+
+Use of this software is governed by the Business Source License included in
+the file licenses/BSL-Couchbase.txt.  As of the Change Date specified in that
+file, in accordance with the Business Source License, use of this software will
+be governed by the Apache License, Version 2.0, included in the file
+licenses/APL2.txt.
+*/
+
+package db
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/couchbase/sync_gateway/base"
+)
+
+const (
+	ClipCapacityHeadroom = 100
+)
+
+type SkippedSequenceListInterface interface {
+	getTimestamp() int64
+	getLastSeq() uint64
+	setLastSeq(seq uint64)
+	getStartSeq() uint64
+	isRange() bool
+}
+
+// SkippedSequenceSlice stores the set of skipped sequences as an ordered slice of single skipped sequences
+// or skipped sequence ranges
+type SkippedSequenceSlice struct {
+	list []SkippedSequenceListInterface
+	lock sync.RWMutex
+}
```

Comment:
Can we implement the following methods (that are defined for SkippedSequenceList today) on SkippedSequenceSlice in this PR, so that we can ensure the implementation is valid for these?
  - getOldest
  - Contains
  - getOlderThan

Reply:
I have changed the IsSkipped method to be named Contains, implemented getOldest with a test. With regards to getOlderThan. My implementation of SkippedSequenceCompact will cover this, my thinking was when we integrate this into the cache we can just call SkippedSequenceCompact from CleanSkippedSequenceQueue. Although getOlderThan returns the sequences its older than a threshold, these sequences are just used to actually go ahead and remove them from the list a little further down in CleanSkippedSequenceQueue (which we already do in SkippedSequenceCompact). So overall my thinking it would just be best to call SkippedSequenceCompact and have it return how many its removing for logging/stats purposes in CleanSkippedSequenceQueue. It will also simplify CleanSkippedSequenceQueue method.